signature/src/policy.rs

```diff
@@ -0,0 +1,460 @@
+// Copyright (c) 2021 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use anyhow::{anyhow, Result};
+use oci_distribution::Reference;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::fs;
+use std::vec::Vec;
+use std::str::FromStr;
+
+use crate::image;
+use crate::signatures;
+
+#[derive(EnumString, Display, Debug, PartialEq)]
+pub enum PolicyReqType {
+    #[strum(to_string = "insecureAcceptAnything")]
+    TypeAccept,
+    #[strum(to_string = "reject")]
+    TypeReject,
+    #[strum(to_string = "signedBy")]
+    TypeSignedBy,
+}
+
+#[derive(EnumString, Display, Debug, PartialEq)]
+pub enum KeyType {
+    #[strum(to_string = "GPGKeys")]
+    Gpg,
+}
+
+#[derive(EnumString, Display, Debug, PartialEq)]
+pub enum PolicyReqMatchType {
+    #[strum(serialize = "matchExact")]
+    TypeMatchExact,
+    #[strum(serialize = "matchRepoDigestOrExact")]
+    TypeMatchRepoDigestOrExact,
+    #[strum(serialize = "matchRepository")]
+    TypeMatchRepository,
+    #[strum(serialize = "exactReference")]
+    TypeExactReference,
+    #[strum(serialize = "exactRepository")]
+    TypeExactRepository,
+    #[strum(serialize = "remapIdentity")]
+    TypeRemapIdentity,
+}
+
+#[derive(EnumString, Display, Debug, PartialEq)]
+pub enum ErrorInfo {
+    #[strum(to_string = "Match reference failed.")]
+    ErrMatchReference,
+    #[strum(to_string = "The policy type name is Unknown.")]
+    ErrUnknownPrType,
+    #[strum(to_string = "The reference match policy type name is Unknown.")]
+    ErrUnknownMatchPolicyType,
+}
+
+// PolicyRequirement is a rule which must be satisfied by the image.
+// It has three types:
+//   `insecureAcceptAnything`: skip signature verification, accept the image unconditionally.
+//   `reject`: reject the image directly.
+//   `signedBy`: there must be at least a signature of the image can be verified by the specific key.
+// The first two are represented by PolicyReqSimple, the last is represented by PolicyReqSignedBy.
+#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
+#[serde(untagged)]
+pub enum PolicyRequirement {
+    SignedBy(PolicyReqSignedBy),
+    Simple(PolicyReqSimple),
+}
```

Comment:
I find this `enum` confusing as it's covering 3 cases with only 2 entries.

What I think would make more sense is using an trait instead of an enum

```Rust
pub trait PolicyRequirement {
     is_image_allowed(&self, image: &mut image::Image);
 }
```

And then have 3 structures that implement that trait:

```Rust
pub struct PolicyReqReject{};
pub struct PolicyReqAcceptAnything{};
pub struct PolicyReqSignedBy {
 ....
}
```

You can implement the `TryFrom` trait on the `PolicyRequirement` trait:

```Rust
impl TryFrom<&str> for &dyn PolicyRequirement {
```

Reply:
I tried to do this, but it will cause problems in deserialization. This is because `serde_json` does not support deserializing to trait objects. `PolicyRequirement` needs to deserialize from JSON strings in the policy configuration file. Using `enum` can make `serde_json::from_str` work now.

I have tried to customize the implementation of the `serde::Deserialize` macro for our `PolicyRequirement` trait object, but it would be very complex and make our code difficult to understand.

Therefore, I think we can use enum here. Maybe serde has better support in the future, and we can consider optimizing the code here.